target/linux/realtek/files-6.6/drivers/thermal/realtek-thermal.c

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Realtek thermal sensor driver
+ *
+ * Copyright (C) 2025 Bjørn Mork <bjorn@mork.no>>
+ */
+
+#include <linux/bitfield.h>
+#include <linux/mfd/syscon.h>
+#include <linux/module.h>
+#include <linux/platform_device.h>
+#include <linux/regmap.h>
+#include "thermal_hwmon.h"
+
+#define RTL8380_THERMAL_METER_CTRL0	0x98
+#define RTL8380_THERMAL_METER_CTRL1	0x9c
+#define RTL8380_THERMAL_METER_CTRL2	0xa0
+#define RTL8380_THERMAL_METER_RESULT	0xa4
+#define RTL8380_TM_ENABLE		BIT(0)
+#define RTL8380_TEMP_VALID		BIT(8)
+#define RTL8380_TEMP_OUT_MASK		GENMASK(6, 0)
+
+#define RTL9300_THERMAL_METER_CTRL0	0x60
+#define RTL9300_THERMAL_METER_CTRL1	0x64
+#define RTL9300_THERMAL_METER_CTRL2	0x68
+#define RTL9300_THERMAL_METER_RESULT0	0x6c
+#define RTL9300_THERMAL_METER_RESULT1	0x70
+#define RTL9300_TM_ENABLE		BIT(16)
+#define RTL9300_TEMP_VALID		BIT(24)
+#define RTL9300_TEMP_OUT_MASK		GENMASK(23, 16)
+
+#define REALTEK_THERMAL_FLAG_INIT	BIT(0)
+
+struct realtek_thermal_priv {
+	struct regmap *regmap;
+	u32 flags;
+};
+
+static void rtl8380_thermal_init(struct realtek_thermal_priv *priv)
+{
+	if (!regmap_update_bits(priv->regmap, RTL8380_THERMAL_METER_CTRL0, RTL8380_TM_ENABLE, RTL8380_TM_ENABLE))
+		priv->flags |= REALTEK_THERMAL_FLAG_INIT;
+}
+
+static int rtl8380_get_temp(struct thermal_zone_device *tz, int *res)
+{
+	struct realtek_thermal_priv *priv = thermal_zone_device_priv(tz);
+        int offset = thermal_zone_get_offset(tz);
+        int slope = thermal_zone_get_slope(tz);
+	u32 val;
+	int ret;
+
+	if (!(priv->flags & REALTEK_THERMAL_FLAG_INIT))
+		rtl8380_thermal_init(priv);
+
+        ret = regmap_read(priv->regmap, RTL8380_THERMAL_METER_RESULT, &val);
+        if (ret)
+                return ret;
+
+	if (!(val & RTL8380_TEMP_VALID))
+		return -EAGAIN;
+
+	*res = FIELD_GET(RTL8380_TEMP_OUT_MASK, val) * slope + offset;
+	return 0;
+}
+
+static const struct thermal_zone_device_ops rtl8380_ops = {
+	.get_temp = rtl8380_get_temp,
+};
+
+static void rtl9300_thermal_init(struct realtek_thermal_priv *priv)
+{
+	if (!regmap_update_bits(priv->regmap, RTL9300_THERMAL_METER_CTRL2, RTL9300_TM_ENABLE, RTL9300_TM_ENABLE))
+		priv->flags |= REALTEK_THERMAL_FLAG_INIT;
+}
+
+static int rtl9300_get_temp(struct thermal_zone_device *tz, int *res)
+{
+	struct realtek_thermal_priv *priv = thermal_zone_device_priv(tz);
+        int offset = thermal_zone_get_offset(tz);
+        int slope = thermal_zone_get_slope(tz);
+	u32 val;
+	int ret;
+
+	if (!(priv->flags & REALTEK_THERMAL_FLAG_INIT))
+		rtl9300_thermal_init(priv);
+
+        ret = regmap_read(priv->regmap, RTL9300_THERMAL_METER_RESULT0, &val);
+        if (ret)
+                return ret;
+
+	if (!(val & RTL9300_TEMP_VALID))
+		return -EAGAIN;
+
+	*res = FIELD_GET(RTL9300_TEMP_OUT_MASK, val) * slope + offset;
+	return 0;
+}
```

Comment:
It looks like this could be deduplicated with rtl8380_get_temp, but you would then need an init op, base register, valid mask, result mask. The latter two could be replaced by a read_result op. In any case it probably wouldn't gain us much at this point. So up to you if you want to investigate.

Reply:
Yes, I already thought about that.  Duplicating code makes me feel sick. But with every constant being different between the two SoCs I believe this is actually the cleanest/smallest/most efficient version in the end.